# Cafe Sales Data Cleaning

## Oasis Infobyte - Data Analytics Internship
### Level 1 - Task 3: Cleaning Data

### Objective

The objective of this project is to systematically clean a messy cafe sales dataset and transform it into a reliable, consistent, and analysis-ready dataset.

The project focuses on:

- Assessing overall data quality
- Identifying and handling missing values
- Detecting duplicate records
- Identifying invalid and inconsistent values
- Standardizing categorical data
- Correcting data types
- Cleaning transaction dates
- Validating numerical relationships
- Detecting and handling outliers
- Comparing data quality before and after cleaning
- Exporting the final cleaned dataset

In [2]:
import pandas as pd
import numpy as np


In [3]:
# Load the dataset
df = pd.read_csv(r"C:\Desktop\OIBSIP\DataAnalytics_L1_DataCleaning\Dataset\Cafe_sales.csv")

# Display first 5 rows
df.head()

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2,4,Credit Card,Takeaway,08-09-2023
1,TXN_4977031,Cake,4,3,12,Cash,In-store,16-05-2023
2,TXN_4271903,Cookie,4,1,ERROR,Credit Card,In-store,19-07-2023
3,TXN_7034554,Salad,2,5,10,UNKNOWN,UNKNOWN,27-04-2023
4,TXN_3160411,Coffee,2,2,4,Digital Wallet,In-store,11-06-2023


## Initial Data Quality Assessment

Before cleaning the dataset, an initial assessment is performed to understand its size, structure, column names, and overall data quality.

In [5]:
print("Dataset Shape:", df.shape)

print("\nColumn Names:")
print(df.columns.tolist())

Dataset Shape: (10020, 8)

Column Names:
['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent', 'Payment Method', 'Location', 'Transaction Date']


### Dataset Information and Data Types

The dataset structure and data types are examined to identify columns with missing values and columns that may require data type conversion during the cleaning process.

In [7]:
# Display dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10020 entries, 0 to 10019
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10020 non-null  object
 1   Item              9686 non-null   object
 2   Quantity          9882 non-null   object
 3   Price Per Unit    9841 non-null   object
 4   Total Spent       9847 non-null   object
 5   Payment Method    7436 non-null   object
 6   Location          6753 non-null   object
 7   Transaction Date  9861 non-null   object
dtypes: object(8)
memory usage: 626.4+ KB


### Missing Values Analysis

Missing values are identified for each column to measure the extent of incomplete data before cleaning. Both the number and percentage of missing values are calculated.

In [9]:
# Calculate missing values and percentages
missing_report = pd.DataFrame({
    'Missing Values': df.isnull().sum(),
    'Missing Percentage (%)': (df.isnull().sum() / len(df) * 100).round(2)
})

missing_report

,Missing Values,Missing Percentage (%)
Transaction ID,0,0.00
Item,334,3.33
Quantity,138,1.38
Price Per Unit,179,1.79
Total Spent,173,1.73
Payment Method,2584,25.79
Location,3267,32.60
Transaction Date,159,1.59


### Duplicate Records

The dataset is checked for exact duplicate rows. Duplicate records can lead to inaccurate calculations and biased analysis, so they should be identified before cleaning.

In [11]:
# Check for exact duplicate rows
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 20


### Duplicate Removal

The dataset contained 20 exact duplicate records. These duplicate rows were removed to prevent repeated transactions from affecting the accuracy of subsequent analysis.

After duplicate removal, the dataset should return to 10,000 unique transaction records.

In [13]:
# Remove duplicate records
rows_before = len(df)

df = df.drop_duplicates().reset_index(drop=True)

rows_after = len(df)
duplicates_removed = rows_before - rows_after

print("Rows before duplicate removal:", rows_before)
print("Duplicates removed:", duplicates_removed)
print("Rows after duplicate removal:", rows_after)

Rows before duplicate removal: 10020
Duplicates removed: 20
Rows after duplicate removal: 10000


### Invalid and Inconsistent Values

In addition to actual missing values, the dataset may contain placeholder values such as `ERROR` and `UNKNOWN`. These values represent invalid or unavailable information and need to be identified before deciding how to handle them.

In [15]:
# Check missing values
print("Missing Values:")
print(df.isnull().sum())

# Check ERROR values
print("\nERROR Values:")
print((df == 'ERROR').sum())

# Check UNKNOWN values
print("\nUNKNOWN Values:")
print((df == 'UNKNOWN').sum())

Missing Values:
Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

ERROR Values:
Transaction ID        0
Item                292
Quantity            170
Price Per Unit      190
Total Spent         164
Payment Method      306
Location            358
Transaction Date    142
dtype: int64

UNKNOWN Values:
Transaction ID        0
Item                344
Quantity            171
Price Per Unit      164
Total Spent         165
Payment Method      293
Location            338
Transaction Date    159
dtype: int64


## Data Cleaning

### Handling Invalid Values

The dataset contains invalid placeholder values such as `ERROR` and `UNKNOWN`. These values are converted to missing values (`NaN`) so that appropriate cleaning strategies can be applied to each column.

In [17]:
# Replace invalid values with NaN
df.replace(['ERROR', 'UNKNOWN'], np.nan, inplace=True)

# Check missing values after replacement
df.isnull().sum()

Transaction ID         0
Item                 969
Quantity             479
Price Per Unit       533
Total Spent          502
Payment Method      3178
Location            3961
Transaction Date     460
dtype: int64

In [18]:
df

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2,4,Credit Card,Takeaway,08-09-2023
1,TXN_4977031,Cake,4,3,12,Cash,In-store,16-05-2023
2,TXN_4271903,Cookie,4,1,NaN,Credit Card,In-store,19-07-2023
3,TXN_7034554,Salad,2,5,10,NaN,NaN,27-04-2023
4,TXN_3160411,Coffee,2,2,4,Digital Wallet,In-store,11-06-2023
...,...,...,...,...,...,...,...,...
9995,TXN_7672686,Coffee,2,2,4,NaN,NaN,30-08-2023
9996,TXN_9659401,NaN,3,NaN,3,Digital Wallet,NaN,02-06-2023
9997,TXN_5255387,Coffee,4,2,8,Digital Wallet,NaN,02-03-2023
9998,TXN_7695629,Cookie,3,NaN,3,Digital Wallet,NaN,02-12-2023


### Correcting Numeric Data Types

The Quantity, Price Per Unit, and Total Spent columns were stored as object data types because they originally contained invalid text values. After replacing these invalid values with missing values, the columns are converted to numeric format.

In [20]:
# Convert columns to numeric
df['Quantity'] = pd.to_numeric(df['Quantity'], errors='coerce')
df['Price Per Unit'] = pd.to_numeric(df['Price Per Unit'], errors='coerce')
df['Total Spent'] = pd.to_numeric(df['Total Spent'], errors='coerce')

# Check data types
df.dtypes

Transaction ID       object
Item                 object
Quantity            float64
Price Per Unit      float64
Total Spent         float64
Payment Method       object
Location             object
Transaction Date     object
dtype: object

The Transaction Date column was stored as an object data type. Since it represents dates, it is converted to datetime format for accurate date-based analysis.

In [22]:
# Convert Transaction Date to datetime
df['Transaction Date'] = pd.to_datetime(
    df['Transaction Date'],
    format='%d-%m-%Y',
    errors='coerce'
)

df['Transaction Date'].head()

0   2023-09-08
1   2023-05-16
2   2023-07-19
3   2023-04-27
4   2023-06-11
Name: Transaction Date, dtype: datetime64[ns]

## Handling Missing Values

### Missing Item Values

Before filling missing values in the Item column, the relationship between Item and Price Per Unit is examined. If individual items have consistent prices, the price information may help identify missing item names.

In [24]:
# Check unique prices for each item
df.groupby('Item')['Price Per Unit'].unique()

Item
Cake        [3.0, nan]
Coffee      [2.0, nan]
Cookie      [1.0, nan]
Juice       [3.0, nan]
Salad       [5.0, nan]
Sandwich    [4.0, nan]
Smoothie    [4.0, nan]
Tea         [1.5, nan]
Name: Price Per Unit, dtype: object

In [25]:
df.isnull().sum()

Transaction ID         0
Item                 969
Quantity             479
Price Per Unit       533
Total Spent          502
Payment Method      3178
Location            3961
Transaction Date     460
dtype: int64

### Handling Missing Item Values

Missing Item values cannot be reliably inferred because some products share the same unit price. Instead of deleting these transactions or guessing the product, missing Item values are labeled as `Unknown` to preserve the available transaction information.

In [27]:
# Fill missing Item values
df['Item'] = df['Item'].fillna('Unknown')

# Check remaining missing Item values
df['Item'].isnull().sum()

0

In [28]:
# Check missing values in numerical columns
df[['Quantity', 'Price Per Unit', 'Total Spent']].isnull().sum()

Quantity          479
Price Per Unit    533
Total Spent       502
dtype: int64

In [29]:
# Check rows with missing numerical values
df[df[['Quantity', 'Price Per Unit', 'Total Spent']].isnull().any(axis=1)].head(10)

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
2,TXN_4271903,Cookie,4.0,1.0,NaN,Credit Card,In-store,2023-07-19
20,TXN_3522028,Smoothie,NaN,4.0,20.0,Cash,In-store,2023-04-04
25,TXN_7958992,Smoothie,3.0,4.0,NaN,NaN,NaN,2023-12-13
31,TXN_8927252,Unknown,2.0,1.0,NaN,Credit Card,NaN,2023-11-06
42,TXN_6650263,Tea,2.0,1.5,NaN,NaN,Takeaway,2023-01-10
55,TXN_5522862,Cookie,NaN,1.0,2.0,Credit Card,Takeaway,2023-03-19
56,TXN_3578141,Cake,5.0,NaN,15.0,NaN,Takeaway,2023-06-27
57,TXN_2080895,Cake,NaN,3.0,3.0,Digital Wallet,In-store,2023-04-19
65,TXN_4987129,Sandwich,3.0,NaN,NaN,NaN,In-store,2023-10-20
66,TXN_8501819,Juice,NaN,3.0,6.0,Cash,NaN,2023-03-30


### Recovering Missing Numerical Values

The numerical columns are logically related:

**Total Spent = Quantity × Price Per Unit**

Therefore, missing numerical values are recovered using the available values wherever possible instead of immediately applying statistical imputation.

In [31]:
# Fill missing Total Spent
df['Total Spent'] = df['Total Spent'].fillna(
    df['Quantity'] * df['Price Per Unit']
)

# Check remaining missing values
df['Total Spent'].isnull().sum()

40

In [32]:
# Fill missing Quantity
df['Quantity'] = df['Quantity'].fillna(
    df['Total Spent'] / df['Price Per Unit']
)

# Check remaining missing values
df['Quantity'].isnull().sum()

38

In [33]:
# Fill missing Price Per Unit
df['Price Per Unit'] = df['Price Per Unit'].fillna(
    df['Total Spent'] / df['Quantity']
)

# Check remaining missing values
df['Price Per Unit'].isnull().sum()

38

In [34]:
# View remaining rows with missing numerical values
df[df[['Quantity', 'Price Per Unit', 'Total Spent']].isnull().any(axis=1)].head(20)

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
65,TXN_4987129,Sandwich,3.0,NaN,NaN,NaN,In-store,2023-10-20
236,TXN_8562645,Salad,NaN,5.0,NaN,NaN,In-store,2023-05-18
278,TXN_3229409,Juice,NaN,3.0,NaN,Cash,Takeaway,2023-04-15
629,TXN_9289174,Cake,NaN,NaN,12.0,Digital Wallet,In-store,2023-12-30
641,TXN_2962976,Juice,NaN,3.0,NaN,NaN,NaN,2023-03-17
738,TXN_8696094,Sandwich,NaN,4.0,NaN,NaN,Takeaway,2023-05-14
912,TXN_1575608,Sandwich,NaN,NaN,20.0,NaN,Takeaway,2023-01-05
1008,TXN_7225428,Tea,NaN,NaN,3.0,Credit Card,Takeaway,2023-03-07
1436,TXN_7590801,Tea,NaN,NaN,6.0,Cash,Takeaway,NaT
1482,TXN_3593060,Smoothie,NaN,NaN,16.0,Cash,NaN,2023-03-05


### Recovering Price Per Unit from Item

Each known cafe item has a fixed unit price. Therefore, remaining missing Price Per Unit values are recovered using the corresponding item's standard price wherever the item name is available.

In [36]:
# Standard price for each item
item_prices = {
    'Cookie': 1.0,
    'Tea': 1.5,
    'Coffee': 2.0,
    'Cake': 3.0,
    'Juice': 3.0,
    'Sandwich': 4.0,
    'Smoothie': 4.0,
    'Salad': 5.0
}

# Fill missing Price Per Unit using Item
df['Price Per Unit'] = df['Price Per Unit'].fillna(
    df['Item'].map(item_prices)
)

# Check remaining missing values
df['Price Per Unit'].isnull().sum()

6

In [37]:
# Fill remaining Quantity where possible
df['Quantity'] = df['Quantity'].fillna(
    df['Total Spent'] / df['Price Per Unit']
)

# Check remaining missing Quantity
df['Quantity'].isnull().sum()

23

In [38]:
# Fill remaining Total Spent where possible
df['Total Spent'] = df['Total Spent'].fillna(
    df['Quantity'] * df['Price Per Unit']
)

# Check remaining missing Total Spent
df['Total Spent'].isnull().sum()

23

In [39]:
# View remaining incomplete numerical records
df[df[['Quantity', 'Price Per Unit', 'Total Spent']].isnull().any(axis=1)]

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
236,TXN_8562645,Salad,NaN,5.0,NaN,NaN,In-store,2023-05-18
278,TXN_3229409,Juice,NaN,3.0,NaN,Cash,Takeaway,2023-04-15
641,TXN_2962976,Juice,NaN,3.0,NaN,NaN,NaN,2023-03-17
738,TXN_8696094,Sandwich,NaN,4.0,NaN,NaN,Takeaway,2023-05-14
1761,TXN_3611851,Unknown,4.0,NaN,NaN,Credit Card,NaN,2023-02-09
2289,TXN_7524977,Unknown,4.0,NaN,NaN,NaN,NaN,2023-12-09
2796,TXN_9188692,Cake,NaN,3.0,NaN,Credit Card,NaN,2023-12-01
3203,TXN_4565754,Smoothie,NaN,4.0,NaN,Digital Wallet,Takeaway,2023-10-06
3224,TXN_6297232,Coffee,NaN,2.0,NaN,NaN,NaN,2023-04-07
3401,TXN_3251829,Tea,NaN,1.5,NaN,Digital Wallet,In-store,2023-07-25


### Removing Remaining Incomplete Numerical Records

Most missing numerical values were recovered using the relationship between Quantity, Price Per Unit, Total Spent, and Item.

However, a small number of records still contained insufficient information to reliably reconstruct the transaction. Since these records represent only a very small proportion of the dataset, they were removed rather than imputing potentially inaccurate numerical values.

In [41]:
# Remove rows with remaining missing numerical values
df = df.dropna(
    subset=['Quantity', 'Price Per Unit', 'Total Spent']
).reset_index(drop=True)

# Check remaining missing numerical values
df[['Quantity', 'Price Per Unit', 'Total Spent']].isnull().sum()

Quantity          0
Price Per Unit    0
Total Spent       0
dtype: int64

### Handling Missing Payment Method

Payment Method contains a large number of missing values. Since payment method is a categorical attribute and the remaining transaction information is still useful, deleting these records would result in unnecessary data loss.

Therefore, missing Payment Method values are labeled as `Unknown`.

In [43]:
# Fill missing Payment Method
df['Payment Method'] = df['Payment Method'].fillna('Unknown')

# Check missing values
df['Payment Method'].isnull().sum()


0

### Handling Missing Location

Location contains a large number of missing values. Since the transaction records still contain useful sales information, removing these rows would cause significant data loss.

Therefore, missing Location values are labeled as `Unknown`.

In [45]:
# Fill missing Location
df['Location'] = df['Location'].fillna('Unknown')

# Check missing values
df['Location'].isnull().sum()

0

### Handling Missing Transaction Dates

Some Transaction Date values remained missing after invalid date entries were converted to missing values.

Since the actual transaction date cannot be reliably inferred from the other attributes, these records are removed rather than assigning artificial dates that could distort time-based analysis.

In [47]:
# Remove rows with missing Transaction Date
df = df.dropna(subset=['Transaction Date']).reset_index(drop=True)

# Check missing Transaction Date values
df['Transaction Date'].isnull().sum()

0

In [48]:
df.shape

(9514, 8)

### Missing Values After Treatment

After applying appropriate cleaning strategies to each column, the dataset is checked again to confirm that all missing values have been handled.

In [50]:
# Check remaining missing values
df.isnull().sum()

Transaction ID      0
Item                0
Quantity            0
Price Per Unit      0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
dtype: int64

In [51]:
print("Item:")
print(df['Item'].unique())

print("\nPayment Method:")
print(df['Payment Method'].unique())

print("\nLocation:")
print(df['Location'].unique())

Item:
['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'Unknown' 'Sandwich' 'Juice'
 'Tea']

Payment Method:
['Credit Card' 'Cash' 'Unknown' 'Digital Wallet']

Location:
['Takeaway' 'In-store' 'Unknown']


### Categorical Data Standardization

The unique values in the Item, Payment Method, and Location columns were examined for inconsistent spelling, capitalization, and formatting.

After replacing invalid values, the remaining categories were found to be consistently formatted. Missing categorical information was represented as `Unknown`, so no further standardization was required.

## Outlier Detection and Treatment

Outliers are examined using the Interquartile Range (IQR) method. This method identifies values that fall significantly below or above the typical range of the data.

The Quantity column is examined first to determine whether unusually high or low transaction quantities are present.

In [54]:
# Calculate IQR for Quantity
Q1 = df['Quantity'].quantile(0.25)
Q3 = df['Quantity'].quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR

print("Lower Limit:", lower_limit)
print("Upper Limit:", upper_limit)

# Count outliers
quantity_outliers = df[
    (df['Quantity'] < lower_limit) |
    (df['Quantity'] > upper_limit)
]

print("Number of Quantity Outliers:", len(quantity_outliers))

Lower Limit: -1.0
Upper Limit: 7.0
Number of Quantity Outliers: 0


### Quantity Outlier Treatment

Using the IQR method, the lower and upper limits for Quantity were calculated as -1 and 7 respectively.

No Quantity values were identified outside these limits. Therefore, no outlier treatment was required for the Quantity column.

In [56]:
# Calculate IQR for Price Per Unit
Q1 = df['Price Per Unit'].quantile(0.25)
Q3 = df['Price Per Unit'].quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR

print("Lower Limit:", lower_limit)
print("Upper Limit:", upper_limit)

# Count outliers
price_outliers = df[
    (df['Price Per Unit'] < lower_limit) |
    (df['Price Per Unit'] > upper_limit)
]

print("Number of Price Per Unit Outliers:", len(price_outliers))

Lower Limit: -1.0
Upper Limit: 7.0
Number of Price Per Unit Outliers: 0


### Price Per Unit Outlier Treatment

Using the IQR method, the lower and upper limits for Price Per Unit were calculated as -1 and 7 respectively.

No Price Per Unit values were identified outside these limits. Therefore, no outlier treatment was required for this column.

In [58]:
# Calculate IQR for Total Spent
Q1 = df['Total Spent'].quantile(0.25)
Q3 = df['Total Spent'].quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR

print("Lower Limit:", lower_limit)
print("Upper Limit:", upper_limit)

# Count outliers
total_outliers = df[
    (df['Total Spent'] < lower_limit) |
    (df['Total Spent'] > upper_limit)
]

print("Number of Total Spent Outliers:", len(total_outliers))

Lower Limit: -8.0
Upper Limit: 24.0
Number of Total Spent Outliers: 259


In [59]:
# View Total Spent outliers
total_outliers[
    ['Item', 'Quantity', 'Price Per Unit', 'Total Spent']
].head(20)

,Item,Quantity,Price Per Unit,Total Spent
10,Salad,5.0,5.0,25.0
48,Salad,5.0,5.0,25.0
49,Unknown,5.0,5.0,25.0
92,Salad,5.0,5.0,25.0
96,Unknown,5.0,5.0,25.0
142,Salad,5.0,5.0,25.0
149,Salad,5.0,5.0,25.0
167,Salad,5.0,5.0,25.0
203,Salad,5.0,5.0,25.0
308,Salad,5.0,5.0,25.0


### Total Spent Outlier Treatment

The IQR method identified 259 Total Spent values outside the calculated upper limit of 24.

Further inspection showed that these values represent valid high-value transactions. For example, a quantity of 5 units at a price of 5 per unit correctly results in a Total Spent value of 25.

Since these values are logically consistent with the relationship:

**Total Spent = Quantity × Price Per Unit**

the detected outliers were retained rather than removed or capped. Removing them would result in the loss of valid transaction information.

In [61]:
# Check Total Spent calculation
df['Calculated Total'] = df['Quantity'] * df['Price Per Unit']

# Find mismatched records
mismatch = df[df['Total Spent'] != df['Calculated Total']]

print("Number of mismatched records:", len(mismatch))

Number of mismatched records: 0


### Numerical Consistency Validation

The Total Spent column was validated against the relationship:

**Total Spent = Quantity × Price Per Unit**

After cleaning, no mismatched records were found. This confirms that the numerical transaction values are internally consistent.

In [103]:
df.shape

(9514, 9)

In [105]:
# Remove temporary validation column
df.drop(columns=['Calculated Total'], inplace=True)

# Check dataset shape
df.shape

(9514, 8)

## Final Data Quality Assessment

After completing the cleaning process, the dataset is reassessed to verify that missing values, duplicate records, data types, and overall structure are suitable for analysis.

In [108]:
print("Dataset Shape:", df.shape)

print("\nMissing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:", df.duplicated().sum())

print("\nData Types:")
print(df.dtypes)

Dataset Shape: (9514, 8)

Missing Values:
Transaction ID      0
Item                0
Quantity            0
Price Per Unit      0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
dtype: int64

Duplicate Rows: 0

Data Types:
Transaction ID              object
Item                        object
Quantity                   float64
Price Per Unit             float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
dtype: object


In [110]:
# Check if any Quantity values contain decimals
decimal_quantity = (df['Quantity'] % 1 != 0).sum()

print("Decimal Quantity Values:", decimal_quantity)

Decimal Quantity Values: 0


### Final Data Type Correction

The Quantity column was stored as a floating-point data type after missing-value treatment. A validation check confirmed that all Quantity values are whole numbers with no decimal values.

Therefore, Quantity was converted to an integer data type to accurately represent the number of items purchased.

In [113]:
df['Quantity'] = df['Quantity'].astype(int)

df.dtypes

Transaction ID              object
Item                        object
Quantity                     int32
Price Per Unit             float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
dtype: object

## Before vs After Data Quality Summary

A comparison is performed between the original working dataset and the final cleaned dataset to demonstrate the improvements achieved through the data cleaning process.

In [116]:
# Load original working dataset for comparison
df_original = pd.read_csv(r"C:\Desktop\OIBSIP\DataAnalytics_L1_DataCleaning\Dataset\Cafe_sales.csv")

summary = pd.DataFrame({
    'Metric': ['Row Count', 'Missing Values', 'Duplicate Rows'],
    'Before Cleaning': [
        len(df_original),
        df_original.isnull().sum().sum(),
        df_original.duplicated().sum()
    ],
    'After Cleaning': [
        len(df),
        df.isnull().sum().sum(),
        df.duplicated().sum()
    ]
})

summary

,Metric,Before Cleaning,After Cleaning
0,Row Count,10020,9514
1,Missing Values,6834,0
2,Duplicate Rows,20,0


### Invalid Placeholder Comparison

In addition to actual missing values, the original dataset contained invalid placeholder values such as `ERROR` and `UNKNOWN`.

These placeholders were identified separately because they are stored as text and are not automatically counted as missing values by pandas.

In [119]:
print("Before Cleaning:")
print("ERROR values:", (df_original == 'ERROR').sum().sum())
print("UNKNOWN values:", (df_original == 'UNKNOWN').sum().sum())

print("\nAfter Cleaning:")
print("ERROR values:", (df == 'ERROR').sum().sum())
print("UNKNOWN values:", (df == 'UNKNOWN').sum().sum())

Before Cleaning:
ERROR values: 1627
UNKNOWN values: 1638

After Cleaning:
ERROR values: 0
UNKNOWN values: 0


### Data Quality Improvement Summary

The cleaning process significantly improved the overall quality of the dataset.

- Row count was reduced from **10,020 to 9,514** after removing duplicate records and rows with insufficient information.
- Missing values were reduced from **6,834 to 0**.
- Duplicate rows were reduced from **20 to 0**.
- `ERROR` placeholder values were reduced from **1,627 to 0**.
- `UNKNOWN` placeholder values were reduced from **1,638 to 0**.
- Numerical columns were converted to appropriate numeric data types.
- Transaction Date was converted to datetime format.
- Numerical relationships were validated for consistency.
- Valid high-value transactions identified by the IQR method were retained.

The final dataset is clean, consistent, and suitable for further analysis.

## 7. Exporting the Cleaned Dataset

After completing all cleaning and validation steps, the final analysis-ready dataset is exported as a new CSV file. The original dataset is retained separately for reference.

In [135]:
# Save final cleaned dataset
df.to_csv(r"C:/Desktop/OIBSIP/DataAnalytics_L1_DataCleaning/Outputs/cleaned_cafe_sales.csv", index=False)

print("Cleaned dataset saved successfully!")
print("Final Dataset Shape:", df.shape)

Cleaned dataset saved successfully!
Final Dataset Shape: (9514, 8)


## Conclusion

The Cafe Sales dataset was successfully cleaned and transformed into an analysis-ready dataset.

The cleaning process included identifying and removing duplicate records, handling missing and invalid values, correcting data types, recovering numerical values using logical relationships, standardizing categorical information, converting transaction dates, detecting outliers using the IQR method, and validating transaction amounts.

The dataset was reduced from **10,020 rows to 9,514 rows**, while missing values were reduced from **6,834 to 0** and duplicate records from **20 to 0**. Invalid `ERROR` and `UNKNOWN` placeholders were also successfully handled.

Outlier analysis showed that some high Total Spent values represented legitimate transactions and were therefore retained. Final validation confirmed that all transaction amounts satisfy the relationship:

**Total Spent = Quantity × Price Per Unit**

The final cleaned dataset contains **9,514 records and 8 columns**, with no missing values or duplicate records, and is ready for further analysis.